### PR TITLE
wazevo(amd64): lowering for div/rem

### PR DESCRIPTION
**Pull request by evacchi**
Not finished yet, I must more tests to verify all the code paths.

### CODE REVIEW

File: internal/engine/wazevo/backend/isa/amd64/machine.go [hunk 1]
```diff
@@ -621,8 +626,6 @@ func (m *machine) lowerShiftR(si *ssa.Instruction, op shiftROp) {
 		m.insert(alu)
 
 	} else {
-		println("::::::SONO QUI")
```
Comment (evacchi) [inline internal/engine/wazevo/backend/isa/amd64/machine.go:624]: leftover garbage...

**Comment by mathetake** [comment]
maybe you might want to do rebase as I merged some bugs

File: internal/engine/wazevo/backend/isa/amd64/machine.go [hunk 2]
```diff
@@ -964,3 +978,127 @@ func (m *machine) frameSize() int64 {
...
+	// If we are compiling a Rem instruction, when the divisor is -1 we land at the end of the function.
+	if ifRemNeg1 != nil {
+		nop, end := m.allocateBrTarget()
+		m.insert(nop)
+		ifRemNeg1.asJmpIf(condZ, newOperandLabel(end))
+	}
+
+	// Restore rax, rdx.
+	m.copyTo(raxTmp, raxVReg)
+	m.copyTo(rdxTmp, rdxVReg)
```
Comment (mathetake) [inline internal/engine/wazevo/backend/isa/amd64/machine.go:1103]: please think deeply about why do you think you need these restore?

File: internal/engine/wazevo/backend/isa/amd64/machine.go [hunk 3]
```diff
@@ -940,8 +945,17 @@ func (m *machine) copyTo(src regalloc.VReg, dst regalloc.VReg) {
...
+		} else {
+			tmp = m.c.AllocateVReg(ssa.TypeV128)
+		}
+	} else {
+		typ := m.c.TypeOf(v)
+		tmp = m.c.AllocateVReg(typ)
+	}
 	m.copyTo(v, tmp)
 	return tmp
 }
```
Comment (mathetake) [inline internal/engine/wazevo/backend/isa/amd64/machine.go:961]: plase revert this change.

**Comment by evacchi** [comment]
rebased, did a quick+dirty change, will review, add more tests. Promising!

File: internal/engine/wazevo/e2e_test.go [hunk 4]
```diff
@@ -122,6 +122,128 @@ func TestE2E(t *testing.T) {
...
+				},
+				{ // This now succeeds.
+					params: []uint64{1, 1, 21, 0},
+					expErr: "wasm error: integer divide by zero",
+				},
+				//{ // This now fails.
+				//	params: []uint64{0x80000000, 0xffffffff, 1, 1},
+				//	expErr: "wasm error: integer overflow",
+				//},
+				{
```
Comment (evacchi) [inline internal/engine/wazevo/e2e_test.go:220]: it seems like if I have more than one error trap in my function, and the generated code is long enough than the error handers down low are generated incorrectly and they SIGSEGV. In fact, this same e2e test above has DivS,RemS and (rem 21%0) crashed. This test case here (Rems, DivS) does not crash and raises the error correctly, but fails on integer overflow (which was handled correctly in the example above)

**Review by mathetake — APPROVED** [review]
